ext/websocket/server.rs

```diff
@@ -0,0 +1,191 @@
+use std::pin::Pin;
+use std::rc::Rc;
+
+use fastwebsockets::CloseCode;
+use fastwebsockets::FragmentCollector;
+use fastwebsockets::Frame;
+use fastwebsockets::OpCode;
+use fastwebsockets::WebSocket;
+
+pub struct ServerWebSocket {
+  ws: AsyncRefCell<FragmentCollector<Pin<Box<dyn Upgraded>>>>,
+}
+
+impl Resource for ServerWebSocket {
+  fn name(&self) -> Cow<str> {
+    "serverWebSocket".into()
+  }
+}
+pub async fn ws_create_server_stream(
+  state: &Rc<RefCell<OpState>>,
+  transport: Pin<Box<dyn Upgraded>>,
+) -> Result<ResourceId, AnyError> {
+  let mut ws = WebSocket::after_handshake(transport);
+  ws.set_writev(false);
+  ws.set_auto_close(true);
+  ws.set_auto_pong(true);
+
+  let ws_resource = ServerWebSocket {
+    ws: AsyncRefCell::new(FragmentCollector::new(ws)),
+  };
+
+  let resource_table = &mut state.borrow_mut().resource_table;
+  let rid = resource_table.add(ws_resource);
+  Ok(rid)
+}
+
+#[op]
+pub async fn op_server_ws_send_binary(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  data: ZeroCopyBuf,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  ws.write_frame(Frame::new(true, OpCode::Binary, None, data.to_vec()))
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_server_ws_send_text(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  data: String,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  ws.write_frame(Frame::new(true, OpCode::Text, None, data.into_bytes()))
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_server_ws_send(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  value: SendValue,
+) -> Result<(), AnyError> {
+  let msg = match value {
+    SendValue::Text(text) => {
+      Frame::new(true, OpCode::Text, None, text.into_bytes())
+    }
+    SendValue::Binary(buf) => {
+      Frame::new(true, OpCode::Binary, None, buf.to_vec())
+    }
+    SendValue::Pong => Frame::new(true, OpCode::Pong, None, vec![]),
+    SendValue::Ping => Frame::new(true, OpCode::Ping, None, vec![]),
+  };
+
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+
+  ws.write_frame(msg)
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op(deferred)]
+pub async fn op_server_ws_close(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  code: Option<u16>,
+  reason: Option<String>,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
```

Comment:
If (code, reason) are (None, None), this should probably return (1005, "").

Reply:
Body in close frame is optional.

https://www.rfc-editor.org/rfc/rfc6455#section-5.5.1

Section 5.5.1:

> The Close frame MAY contain a body (the "Application data" portion of
   the frame) that indicates a reason for closing

ref: tungstenite https://github.com/snapview/tungstenite-rs/blob/92d65e11043f2a3bf18099ecc3cd7911e70bc456/src/protocol/frame/frame.rs#L347

https://www.rfc-editor.org/rfc/rfc6455#section-7.4

Section 7.4:

> 1005 is a reserved value and MUST NOT be set as a status code in a
      Close control frame by an endpoint.  It is designated for use in
      applications expecting a status code to indicate that no status
      code was actually present.